src/Features/LanguageServer/Protocol/Workspaces/LspWorkspaceManager.cs

```diff
@@ -246,35 +270,43 @@ async Task<(Solution Solution, bool IsForked)> GetLspSolutionForWorkspaceAsync(W
+            //     a reasonable price to pay.  For example, we always do this in VS anyways to make OOP calls, and it is
+            //     not a burden there.
             //
-            //   2.  If the cached solution isn't a match, we compare the LSP text to the workspace's text and return
-            //       the workspace text if all LSP text matches. This check is performant as checksums will be computed
-            //       for these documents in order to make requests OOP.  So these are already computed or will be later
-            //       in this request.
+            //  4. Third, we check to see if we have cached a forked LSP solution for the current set of LSP texts
+            //     against the current workspace version. If so, we can just reuse that instead of re-forking and
+            //     blowing away the trees / source generated docs / etc. that we created for the fork.
             //
-            //   3.  Third, we check to see if we have cached a forked LSP solution for the current set of LSP texts
-            //       against the current workspace version. If so, we can just reuse that instead of re-forking and
-            //       blowing away the trees / source generated docs / etc. that we created for the fork.
-            //       
-            //   4.  We have nothing cached for this combination of LSP texts and workspace version.  We have exhausted
-            //       our options and must create an LSP fork from the current workspace solution with the current LSP
-            //       text.
+            //  5. We have nothing cached for this combination of LSP texts and workspace version.  We have exhausted
+            //     our options and must create an LSP fork from the current workspace solution with the current LSP
+            //     text.
             //
             // We propagate the IsForked value back up so that we only report telemetry on forking if the forked
             // solution is actually requested.
 
             // Step 1: Check if nothing has changed and we already verified that the workspace text matches our LSP text.
             if (_cachedLspSolutions.TryGetValue(workspace, out var cachedSolution) && cachedSolution.solution == workspaceCurrentSolution)
-            {
                 return (workspaceCurrentSolution, IsForked: false);
-            }
 
-            // Step 2: Check to see if the LSP text matches the workspace text.
+            // Step 2: Push through any changes to the underlying workspace if it's a mutating workspace.
+            OpenOrEditDocumentsInMutatingWorkspace(workspace);
+
+            // Because the workspace may have been mutated, go back and retrieve its current snapshot so we're operating
+            // against that view.
+            workspaceCurrentSolution = workspace.CurrentSolution;
+
+            // Step 3: Check to see if the LSP text matches the workspace text.
```

Comment:
(I'm actually more generally wondering why we have this caching logic at all, versus there's just a different implementation of LspWorkspaceManager or something like that.)

Reply:
The update to the workspace could fail or the workspace could change underneath us (e.g. document is removed entirely).  We definitely still need to verify that the workspace text matches and fork if it does not.  Now forking should be rare, but its still required.

Reply:
I agree with David.  A mutating workspace doesn't mean things are always in sync.  The model i'm taking (which i think is actually quite sensible) is that the workspace's *structure* (e.g. set of projects/files) is entirely determined by the project system, on its cadence.  *However*, the LSP server also maintains a view of LSP contents for all files (indicated by 'uri') that is in sync with the *client*. 

Both before and after my change, these two views can be held without problem.  Specifically, the LSP server consults the workspace produced by the project system, but it always exposes a solution-snapshot to features that exactly matches the content of teh LSP client for open files.

None of the above changes with this PR.  All that changes is that instead of having the core-workspace and LSP server potentially disagreeing greatly on the content of open files (necessitating forking), now nearly all of hte time they will agree since the core-workspace contents will be updated (optimistically) to match the LSP contents.

Reply:
> Now forking should be rare, but its still required.

Correct.  We still have the concept that there is no orchestrator ensuring LSP and PS are in agreement.  It's not even clear waht that would mean.  The PS might decide to make changes (like remove files) that the LSP client still thinks exist.  So it's very necessary that while we use the structure provided by teh PS as the right *starting point* we always recognize and reconcile that the files (And contents) the client talks to us about may not be in full alignment.